lib/docker.ml

```diff
@@ -1,37 +1,47 @@
 open Lwt.Infix
 
-let export_env base : Config.env Lwt.t =
-  Os.pread ["docker"; "image"; "inspect";
-            "--format"; {|{{range .Config.Env}}{{print . "\x00"}}{{end}}|};
-            "--"; base] >|= fun env ->
-  String.split_on_char '\x00' env
-  |> List.filter_map (function
-      | "\n" -> None
-      | kv ->
-        match Astring.String.cut ~sep:"=" kv with
-        | None -> Fmt.failwith "Invalid environment in Docker image %S (should be 'K=V')" kv
-        | Some _ as pair -> pair
-    )
+let export_env config : Config.env =
+  Docker_hub.Config.env config |>
+  List.filter_map (fun kv ->
+    match Astring.String.cut ~sep:"=" kv with
+    | None -> Fmt.failwith "Invalid environment in Docker image %S (should be 'K=V')" kv
+    | Some _ as pair -> pair
+  )
 
-let with_container ~log base fn =
-  Os.with_pipe_from_child (fun ~r ~w ->
-      (* We might need to do a pull here, so log the output to show progress. *)
-      let copy = Build_log.copy ~src:r ~dst:log in
-      Os.pread ~stderr:(`FD_move_safely w) ["docker"; "create"; "--"; base] >>= fun cid ->
-      copy >|= fun () ->
-      String.trim cid
-    ) >>= fun cid ->
-  Lwt.finalize
-    (fun () -> fn cid)
-    (fun () -> Os.exec ~stdout:`Dev_null ["docker"; "rm"; "--"; cid])
+let handle_errors = function
+  | Ok x -> Lwt.return x
+  | Error _ -> (* TODO: pretty print the errors *)
+      Lwt.fail_with "TODO"
 
+let with_container manifest token fn =
+  Lwt_io.with_temp_dir ~perm:0o700 ~prefix:"obuilder-docker-hub-" @@ fun output_file ->
+  Docker_hub.fetch_rootfs ~output_file:(Fpath.v output_file) manifest token >>=
+  handle_errors >>= fun () ->
+  fn output_file
```

Comment:
`Lwt_io.with_temp_dir` will fail to delete the directory otherwise
```suggestion
  fn output_file >>= fun () ->
  Os.sudo ["chflags"; "-R"; "0"; output_file]
```